src/tiledb/cloud/files/utils.py

```diff
@@ -264,3 +267,134 @@ def _auth_headers(cfg: configuration.Configuration) -> Mapping[str, object]:
         return {"authorization": basic}
     return {}
     # No authentication has been provided. Do nothing.
+
+
+def upload_folder(
+    input_uri: str,
+    output_uri: str,
+    *,
+    group_uri: Optional[str] = None,
+    exclude_files: Optional[List[str]] = None,
+    flatten: bool = False,
+    access_credentials_name: Optional[str] = None,
+    config: Optional[dict] = None,
+    verbose: bool = False,
+) -> Dict[str, Union[str, Dict[str, str]]]:
+    """
+    Uploads a folder to TileDB Cloud.
+    By default respects the initial folder structure in the destination.
+
+    :param input_uri: The URI or path of the input file. May be an ordinary path
+        or any URI accessible via TileDB VFS.
+    :param output_uri: The TileDB URI to write the file to.
+    :param group_uri: A TileDB Group URI to ingest folder into, defaults to None
+    :param exclude_files: A list of file paths to exclude from uploading,
+        defaults to None.
+    :param flatten: Flag. If set to True, the upload will flatten the folder
+        structure instead of recreating it.
+    :param access_credentials_name: If present, the name of the credentials
+        to use when writing the uploaded file to backend storage instead of
+        the defaults.
+    :param config: Config dictionary, defaults to None
+    :param verbose: Verbose logging, defaults to None
+    :return: A dictionary containing a report message
+        and an upload errors dictionary (if any)
+    """
+    logger = get_logger_wrapper(verbose)
+
+    # Prepare and sanitize arguments
+    output_uri = output_uri.strip("/")
+    exclude_files = exclude_files or []
+
+    input_uri = input_uri if input_uri.endswith(os.sep) else input_uri + os.sep
+    base_dir = os.path.dirname(input_uri)
+    base_dir = os.path.basename(base_dir)
+
+    namespace, name = utils.split_uri(output_uri)
+    _, sp, acn = groups._default_ns_path_cred(namespace=namespace)
+
+    storage_path = name if name.startswith("s3://") else sp
+    storage_path = f"{storage_path.strip('/')}/{base_dir}"
+    logger.debug("Output storage path: %s" % storage_path)
+
+    access_credentials_name = access_credentials_name or acn
+
+    # Group check and/or creation
+    if not group_uri:
+        logger.debug("No group_uri provided. Choosing one...")
+        if not name.startswith("s3://"):
+            group_uri = f"tiledb://{namespace}/{base_dir}"
+        else:
+            group_uri = output_uri
+
+    group_created = False
+    if not tiledb.object_type(group_uri, ctx=tiledb.cloud.Ctx()) == "group":
+        group_namespace, group_name = utils.split_uri(group_uri)
+        groups.create(
```

Comment:
Let's use TileDB-Py library group functionality here. We also need to open the group and add all the members.